### PR TITLE
Metadata

### LINKED ISSUE #118: Support metadata in JSON graphs
Currently MsgFlo drops all metadata from JSON graphs, meaning that we can't give nodes friendly names, edges colors etc.

Now you end up using the role name as the label, which is not so nice:

screenshot 2017-05-05 at 20 19 33

### PR DESCRIPTION
**Pull request by bergie**
Will fix #118

### CODE REVIEW

File: src/coordinator.coffee [hunk 1]
```diff
@@ -537,15 +573,16 @@ class Coordinator extends EventEmitter
...
-          @startParticipant role, componentName, cb
+          metadata = graph.processes[role].metadata or {}
+          @startParticipant role, componentName, metadata, cb
         , setupCallback
 
       setupConnections = (setupCallback) =>
         async.map graph.connections, (c, cb) =>
           if c.data
-            @addInitial c.tgt.process, c.tgt.port, c.data, cb
+            @addInitial c.tgt.process, c.tgt.port, c.data, c.metadata, cb
```
Comment (jonnor) [inline src/coordinator.coffee]: Is `c.metadata` guaranteed to be an object? If it can be null/undefined, should normalize to empty object - either here or inside addInitial (maybe better).

Comment (bergie) [inline src/coordinator.coffee]: Added in 8d73552

File: src/coordinator.coffee [hunk 2]
```diff
@@ -244,6 +249,17 @@ class Coordinator extends EventEmitter
...
 
+  updateParticipantMetadata: (node, metadata, callback) ->
+    metadata = {} unless metadata
+    process = null
+    for k, v of @processes
+      if k == node
+        process = v
+    unless process
+      return callback new Error "Node #{node} not defined"
+    @processes[k].metadata = metadata
```
Comment (jonnor) [inline src/coordinator.coffee]: This is maybe not quite right... `@processes` are the `child_process` instances, for (and only for) the things which the coordinator starts. Things which are discovered don't show up here, but may still have metadata attached? 
Also affects roles which are ignored...

But maybe we cannot do much about this until we have a clearer split of actual/desired states.

Comment (bergie) [inline src/coordinator.coffee]: Yes, the actual/desired split is the next big thing to do

Comment (jonnor) [inline src/coordinator.coffee]: For `@connections` we already have the concept, as we add an entry immediately (with empty queue info).

However, both `@participants` and `@processes` have different shapes than the (desired) FBP graph. There may be multiple participants per role, it may not actually run. Processes is only for participants started by us, not those discovered externally.

So I'm thinking that we (not in this PR) should add a `@roles`, which would track the "nodes" (in FBP). Probably would make the `start/stopParticipant()` API internal, wrapped by a `add/removeRole()`

File: src/coordinator.coffee [hunk 3]
```diff
@@ -244,6 +249,17 @@ class Coordinator extends EventEmitter
         delete @process[k]
       return callback null, processes
 
+  updateParticipantMetadata: (node, metadata, callback) ->
```
Comment (jonnor) [inline src/coordinator.coffee]: The naming is wrong. This is not participant metadata, but role (or node).

File: src/coordinator.coffee [hunk 4]
```diff
@@ -475,6 +514,7 @@ class Coordinator extends EventEmitter
       name = part.role
       graph.processes[name] =
         component: part.component
+        metadata: @participants[id].metadata
```
Comment (jonnor) [inline src/coordinator.coffee]: Wrong, metadata is not stored on `@participants` ?

File: spec/protocol.coffee [hunk 5]
```diff
@@ -331,14 +351,34 @@ describe 'FBP runtime protocol', () ->
...
+      ui.on 'message', checkMessage
+      change =
+        id: 'mycoffeescriptproducer'
+        graph: 'default/main'
+        metadata:
+          label: 'mycoffeeproducer'
+          x: 2
+      ui.send 'graph', 'changenode', change
 
   describe 'subscribing to edges', ->
```
Comment (jonnor) [inline spec/protocol.coffee]: Need a test for whether main graph, as returned by `getsource`, contains the right metadata. That is what Flowhub live-mode will mostly rely on

File: src/coordinator.coffee (hunk 2 above)
Comment (jonnor) [inline src/coordinator.coffee]: Probably then startParticipant would no longer take `metadata` as parameter (as its handled before/outside)

File: src/coordinator.coffee (hunk 4 above)
Comment (bergie) [inline src/coordinator.coffee]: Should we add a new `@nodes` data structure for this stuff, then?

**Comment by bergie** [comment]
@jonnor added the requested test. Storing metadata to `@participants` is not the best option, but I would elect to do it that way in this PR, as otherwise we'll get scope creep towards desired/actual. That should be done next, though.

File: src/coordinator.coffee [hunk 6]
```diff
@@ -244,6 +249,17 @@ class Coordinator extends EventEmitter
         delete @process[k]
       return callback null, processes
 
+  updateNodeMetadata: (node, metadata, callback) ->
+    metadata = {} unless metadata
+    process = null
+    for k, v of @participants
```
Comment (jonnor) [inline src/coordinator.coffee]: I don't get it. `startParticipant` adds metadata to `@processes`, but this updates metadata on `@participants` (but only the last one that matches)?
I don't see how these two data sources are reconsiled...

File: src/coordinator.coffee [hunk 7]
```diff
@@ -97,9 +97,10 @@ pingUrl = (address, method, callback) ->
 
 class Coordinator extends EventEmitter
   constructor: (@broker, @options = {}) ->
-    @participants = {}
+    @participants = {} # participantId -> Definition (from discovery)
     @connections = {} # connId -> { queue: opt String, handler: opt function }
     @iips = {} # iipId -> { metadata, data }
+    @nodes = {} # role -> { metadata: {} }
```
Comment (bergie) [inline src/coordinator.coffee]: We should probably also keep `component` here, as that way the nodes+connections would be a full representation of the desired graph

Comment (bergie) [inline src/coordinator.coffee]: Though, maybe we'll do that in a separate PR where we then make the `nodes` list be what gets dumped to JSON graph instead of participants.

Comment (jonnor) [inline src/coordinator.coffee]: Yes that was my intention. Question remains what to do with discovered things which are not part of `nodes`. They have to be in the graph somehow, otherwise Flowhub will never know about them and it wont be possible to add them?

Comment (bergie) [inline src/coordinator.coffee]: I would just go and add anything discovered to nodes. So then the graph is:

* Things that graph designer wanted to be there (whether they're currently running or not)
* _In addition_ other things that are running

It can be a bit confusing to have the differentiation between:

* Processes you can start and stop by adding/removing them from the graph (things coordinator manages)
* Processes that are essentially out of your control (things coordinator ignores but send discovery)

But that can't be helped. Maybe we can pass some metadata to signify the difference, since for example in the latter case trying to remove such node will be no-op.
